mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala

```diff
@@ -205,6 +227,14 @@ class ImputerModel private[ml] (
 
   import ImputerModel._
 
+  /** @group setParam */
+  @Since("3.0.0")
+  def setInputCol(value: String): this.type = set(inputCol, value)
```

Comment:
What is the intended purpose of this method? 

As it is implemented right now, it doesn't seem to have any practical applications:

- If model has been created with single col, surrogate will contain only a single column, so there is nothing to set here.
- If model has been created with multiple cols, `setInputCol` / `setOutputCol` should clear `setInputCols` and `setOutputCols`, otherwise it will fail to validate.

      @Since("3.0.0")
      def setInputCol(value: String): this.type = {
       clear(inputCols)
       clear(outputCols)
        set(inputCol, value)
      }

      @Since("3.0.0")
      def setOutputCol(value: String): this.type = {
        clear(inputCols)
        clear(outputCols)
         set(outputCol, value)
      }


I am asking, because these two are missing in Python (https://github.com/apache/spark/pull/27195).

Reply:
@zero323 I actually realized I missed the two setters in python when I checked the parity between python and scala last night. I fixed it along with a few other problems.

Reply:
@zero323 There is a check on scala side to make sure only ```setInputCol/setOutputCol``` or ```setInputCols/setOutputCols``` is set

Reply:
> @zero323 There is a check on scala side to make sure only `setInputCol/setOutputCol` or `setInputCols/setOutputCols` is set

That's is what confuses me.  Let's say the workflow looks like this:

```
import org.apache.spark.ml.feature.Imputer

val df = Seq((1, 2)).toDF("x1", "x2")

val mm = new Imputer()
   .setInputCols(Array("x1", "x2"))
   .setOutputCols(Array("x1_", "x2_"))
   .fit(df)
```

You cannot switch to single `col` at the model level:

```
mm.setInputCol("x1").setOutputCol("x1_").transform(df)

// java.lang.IllegalArgumentException: requirement failed: ImputerModel ImputerModel: uid=imputer_5923f59d0d3a, strategy=mean, missingValue=NaN, numInputCols=2, numOutputCols=2 requires exactly one of inputCol, inputCols Params to be set, but both are set.
```

without clearing `cols` explicitly:

```
mm.clear(mm.inputCols).clear(mm.outputCols).transform(df)
```

That's really not intuitive workflow, if this is what was intended.

If we only want to support `Imupter.setInputCol` -> `ImputerModel.setInputcol`, then there is no point in having this method at all:

```
val ms = new Imputer().setInputCol("x1").setOutputCol("x1_").fit(df)

// ms.setInputCol("x2").setOutputCol("x2_").transform(df)
org.apache.spark.sql.AnalysisException: cannot resolve '`x2`' given input columns: [x1];;

```

as surrogate contains only the column used for fit

```
scala> ms.surrogateDF
res13: org.apache.spark.sql.DataFrame = [x1: double]
```

Reply:
@zero323 It is a problem. I will have a follow up pr to fix this.

Reply:
Thanks @huaxingao. Please ping me when you do.

Reply:
I think the sanity check in Scala side for inputCol/outputCol and inputCols/outputCols are more for preventing errors when mixing single and multiple columns at the same time, e.g. set both single and multiple column params, inputCol + outputCols...etc.

It sounds rarely switching between single/multiple column during fitting and transforming.

Reply:
@viirya But then we're back to the question why we need `setInputCol` in  `Models`. Should we support `Estimator.setInputCols(...).fit(...).setInputCol(...).transform(...)` flow at all?

Reply:
And what about overriding `inputCols` (taking a subset?)

    new Estimator().setInputCols(...).fit(...).setInputCols(...).transform(...)

for that matter?

Reply:
For a model fitted by an Estimator, I think we usually won't change input/output column(s). The setter is still useful, as there are still cases that we might create a model instance directly. For such cases, we need input column(s) setter.

Reply:
Having option to overwrite `outputCol(s)` on can be useful to avoid name clashes on pre-trained models. 

But providing setters for inputs seems to be more confusing than useful, and proliferation of `Params` that support both `Col` and `Cols` makes things even more fuzzy.

In general I am asking because we seem to have cases like `OneHotEncoderModel` - which provide `setInputCols` / `setOutputCols` but no single column equivalents.